WIP: Document the design of Nix (#4280)
NixOS/nix: doc/manual/src/design/store/entries.md

# Store Entries

File system data in Nix is organized into *store entries*.
A store entry is the combination of

  - some file system data
  - references to store entries

## File system data

The nix store uses a simple filesystem model, similar to the one Git uses.
In particular, every file system object falls into these three cases:

 - File: arbitrary data

 - Directory: mapping of names to child file system objects.
   File children additionally have an executable flag.

 - Symlink: may point anywhere.
   In particular, Symlinks that do not point within the containing file system data or that of another store entry referenced by the containing store entry are allowed, but might not function as intended.

A bare file as the "root" file system object is allowed.
Note that it cannot be executable, though.
This is a consequence of the executable flags being part of the child entries of the directory, rather than the child files themselves.
A root file has no parent directory; so there is no child entry about the root file, and therefore no executable flag for it.

> Without a flag saying which, whether root files are executable or non-executable by default must be decided by convention, and the choice of Nix (and git) is to make them non-executable.

Author: Fixed by splitting section.

## References

Store entries can refer to both other store entries and themselves.

References are normally calculated by scanning the file system data for store paths (which we describe in the next section) referring to store entries.
For now, it suffices to say that a store path is a string encoding of a reference to a store paths, and therefore it is something that we can search for in the contents of files, and thus in store entries by searching in all their files.   

> When we get to building in a future section, this process will be described in precise detail.

Author: Agreed. Added the reference I plan on making. Reword to emphasize invariants over creation process.

However, scanning for references is not mandatory.
Store entries are allowed to have official references that *don't* correspond to store paths contained in their contents,
and they are also allowed to *not* have references that *do* correspond to store paths contained in their store.
Taken together, this means there is no actual rule relating the store paths contained in the contents to the store paths deemed references.

> This is why it's its necessary for correctness, and not just performance, that Nix remember the references of each store entry, rather than try to recompute them on the fly by scanning their contents.

Author: This gone, moved to the relocation section perhaps.

> # Store Entries

Author: We agreed. "Store objects" it is.

> File system data in Nix is organized into *store entries*.

Author: Agreed "file system objects" is something else, but store objects don't necessary need to represent stuff in a file system. E.g. stores that don't mount can "mix" store objects in different store dirs.

> File system data in Nix is organized into *store entries*.

Author: It is now simply

```suggestion
Data in Nix is chiefly organized into *store objects*.
```

A store entry is the combination of

>   - some file system data

Author: Good point. But It is a single file system object, then, since the object contains/owns its children?

>   - some file system data

Author: Store entry -> Store object then, too.

>   - some file system data

Reviewer: See my question on [store objects](https://github.com/NixOS/nix/pull/4280#discussion_r831163072). I actually do like "store entry", as it made sense to me intuitively. But for keeping consistency I'm all for using "store object". Maybe @edolstra can give us a nod?

  - references to store entries

> ## File system data

Author: Done

> Nix supports the a similar model of the file system as Git.

Reviewer 2: I found this sentence pretty confusing. Maybe something like, "The nix store uses a simple filesystem model, similar to Git: ..."

Author: ```suggestion
The nix store uses a simple filesystem model, similar to the one Git uses.
```

> Nix supports the a similar model of the file system as Git.

Author: It now just says "The Nix store uses a simple filesystem model." Git has some key differences, and so is discussed later.

> Namely, every file system object falls into these three cases:

Author: ```suggestion
In particular, every file system object falls into these three cases:
```

 - File: arbitrary data

 - Directory: mapping of names to child file system objects.
   File children additionally have an executable flag.

>  - Symlink: may point anywhere.

Author: Done.

>    In particular, Symlinks that do not point within the containing file system data or that of another store entry referenced by the containing store entry are allowed, but might not function as intended.

Author: Well keep in mind they could point to an unrelated store entry, but if that is not in the closure, that store entry can be GC'd.

>    In particular, Symlinks that do not point within the containing file system data or that of another store entry referenced by the containing store entry are allowed, but might not function as intended.

Reviewer: How is the closure computed? I thought Nix traverses the references.

>    In particular, Symlinks that do not point within the containing file system data or that of another store entry referenced by the containing store entry are allowed, but might not function as intended.

Author: It is possible to have symlinks that point to a '/nix/store/adsfasdfasd-foo` that is *not* a reference.

I can't just do
```nix
runCommand "sneayk {} "ln -s /nix/store/adsfasdfasd-bill-gates-private-key $out" 
```
and get access to it!

Conversely I could do
```nix
runCommand "sneayk {} "ln -s /nix/store/adsfasdfasd-rando $out" 
```
where `/nix/store/adsfasdfasd-rando` is a non-dep, but one that is already installed. The symlink points to a non-referenced path, but it works when ambiently accessing the store (as opposed to being in a downstream build with sandboxing) only while `/nix/store/adsfasdfasd-rando` happens to be still around.

A bare file as the "root" file system object is allowed.

> Note that there is no containing directory object to store its executable bit; it's deemed non-executable by default.

Reviewer 3: I had some trouble understanding this sentence on my first few passes. How about:

"Note that directories describe if a file is executable or not, therefore when a file has no containing directory, it cannot be executable"

It's also worth mentioning that this is not a default, or non-default behavior. The way it's described in the bullets  above, it is impossible for a bare file as a store path to be executable.

Author: ```suggestion
Note that it cannot be executable, though.
This is a consequence of the executable flags being part of the child entries of the directory, rather than the child files themselves.
A root file has no parent directory; so there is no child entry about the root file, and therefore no executable flag for it.
Without a flag saying which, whether root files are executable or non-executable by default must be decided by convention, and the choice of Nix (and git) is to make them non-executable.
```

> Note that there is no containing directory object to store its executable bit; it's deemed non-executable by default.

Author: Turns out this is just true for `git`, so I moved this all to another "aside" subsection. I hope it is clearer now.

## References

Store entries can refer to both other store entries and themselves.

> Store references are normally calculated by scanning the file system data for store paths when a new store entry is created, but this isn't mandatory, as store entries are allowed to have references that don't correspond to contained store paths, and contained store paths that don't correspond to references.

Reviewer 3: The use of "normally" makes me question if store references may be computed in other ways. I don't think that's the case, so perhaps removing "normally" would help.

Alternatively, "contained" might refer to "contained in the closure of a store path". The way it's used here sort of implies that's how closures are tracked.

> Store references are normally calculated by scanning the file system data for store paths when a new store entry is created, but this isn't mandatory, as store entries are allowed to have references that don't correspond to contained store paths, and contained store paths that don't correspond to references.

Author: ```suggestion
References are normally calculated by scanning the file system data for store paths (which we describe in the next section) referring to store entries.
For now, it suffices to say that a store path is a string encoding of a reference to a store paths, and therefore it is something that we can search for in the contents of files, and thus in store entries by searching in all their files.   
When we get to building in a future section, this process will be described in precise detail.

However, scanning for references is not mandatory.
Store entries are allowed to have official references that *don't* correspond to store paths contained in their contents,
and they are also allowed to *not* have references that *do* correspond to store paths contained in their store.
Taken together, this means there is no actual rule relating the store paths contained in the contents to the store paths deemed references.

This is why it's its necessary for correctness, and not just performance, that Nix remember the references of each store entry, rather than try to recompute them on the fly by scanning their contents.
```

> Store references are normally calculated by scanning the file system data for store paths when a new store entry is created, but this isn't mandatory, as store entries are allowed to have references that don't correspond to contained store paths, and contained store paths that don't correspond to references.

Author: Yes, made that more explicit.

> Store references are normally calculated by scanning the file system data for store paths when a new store entry is created, but this isn't mandatory, as store entries are allowed to have references that don't correspond to contained store paths, and contained store paths that don't correspond to references.

Author: Rewrote this already, most complexity saved for later.

The references themselves need not be store paths per-se (this is an implementation detail of the store).
But, like rendered store paths (see next section) in addition to identifying store entries they must also identify the store directory of the store(s) that contain those store entries.
That said, all the references of the store entry must agree on a store dir.

> Also the store directory of the references must equal that of any store which contains the store entry doing the referencing.

Reviewer 3: I also have trouble with this paragraph. I think it's trying to say that references:
 * identify a store entry. 
 * don't need to be store paths.
 * must have the same value across all stores. 

(not sure about this last part, it is really hard for me to understand that last sentence)

> Also the store directory of the references must equal that of any store which contains the store entry doing the referencing.

Author: It might be best to delay talking about this until later.

> Also the store directory of the references must equal that of any store which contains the store entry doing the referencing.

Author: Saving this for a later section. Hopefully what is have now is simpler by omitting these out-of-scope details, and thus easier to understand.

> ## Relocatability

Author: Done in ae23bf3898c6dfa6f72c3c2a0b68b2461bd4bf67

> The two final restrictions of the previous section yield an alternative of view of the same information.

Reviewer 3: which "of" should be removed? I think the sentence reads fine as is.

> The two final restrictions of the previous section yield an alternative of view of the same information.

Reviewer 4: Was thinking "alternative of view of" => "alternative view of". (Non native english speaker here, so maybe just ignore that comment)

> The two final restrictions of the previous section yield an alternative of view of the same information.

Reviewer 3: Ah, yes that's a good catch. I read it like 5 times, and missed that "of" every time.

> The two final restrictions of the previous section yield an alternative of view of the same information.

Author: ```suggestion
The two final restrictions of the previous section yield an alternative view of the same information.
```

> Rather than associating store dirs with the references, we can say a store entry itself has a store dir if and only if it has at least once reference.

Author: ```suggestion
Rather than associating store dirs with the references, we can say a store entry itself has a store dir if and only if it has at least one reference.
```

This corresponds to the observation that a store entry with references, i.e. with a store directory under this interpretation, is confined to stores sharing that same store directory, but a store entry without any references, i.e. thus without a store directory, can exist in any store.

Lastly, this illustrates the purpose of tracking self references.
Store entries without self-references or other references are relocatable, while store paths with self-references aren't.
This is used to tell apart e.g. source code which can be stored anywhere, and pesky non-reloctable executables which assume they are installed to a certain path.

> \[The default method of calculating references by scanning for store paths handles these two example cases surprisingly well.\]

Reviewer 4: There are quite a few occurrences of the word "store" in these sentences to the point where it gets confusing. Not sure how to resolve that though.

Author: This whole section is too advanced, so I moved it to a bonus topic at the end.

> \[The default method of calculating references by scanning for store paths handles these two example cases surprisingly well.\]

Author: Saved for `doc/manual/src/design/store/relocatability.md`
